File: src/todo_sync/things_reader.py
"""Read tasks from Things 3 via the things.py library."""

from __future__ import annotations

from datetime import datetime

from todo_sync.taskwarrior import build_task_data


def fetch_things_tasks(config: dict) -> list[dict]:
    """Fetch incomplete tasks from Things 3 and return normalized task dicts."""
    import things

    things_cfg = config.get("things", {})
    sync_cfg = config.get("sync", {})
    source_tag = sync_cfg.get("things_tag", "things3")
    area_filter = things_cfg.get("areas", [])

    raw_tasks = things.todos()

    # If area filter is set, look up area titles and filter
    if area_filter:
        areas = {a["uuid"]: a["title"] for a in things.areas()}
        raw_tasks = [
            t for t in raw_tasks
            if t.get("area") and areas.get(t["area"]) in area_filter
        ]

    # Look up area and project titles for mapping
    areas = {a["uuid"]: a["title"] for a in things.areas()}
    projects = {p["uuid"]: p["title"] for p in things.projects()}

    result = []
    for t in raw_tasks:
        # Resolve project name
        project_name = None
        if t.get("project"):
            project_name = projects.get(t["project"])
        elif t.get("area"):
            # Use area as project if no project is set
            project_name = areas.get(t["area"])

        # Parse deadline
        due = None
        if t.get("deadline"):
            try:
                due = datetime.strptime(t["deadline"], "%Y-%m-%d")
            except (ValueError, TypeError):
                pass

        # Collect tags
        tags = list(t.get("tags", []) or [])

        # Build annotations from notes
        annotations = []
        if t.get("notes"):
            annotations.append(t["notes"])

        task_data = build_task_data(
            description=t["title"],
            external_id_field="things3_uuid",
            external_id=t["uuid"],
            source="things3",
            source_tag=source_tag,
            project=project_name,
            due=due,
            tags=tags,
            annotations=annotations,
        )
        result.append(task_data)

    return result

File: ts/src/things-reader.ts
/**
 * Read tasks from Things 3 via its SQLite database.
 *
 * Things 3 stores its data in:
 *   ~/Library/Group Containers/JLMPQHK86H.com.culturedcode.ThingsMac/
 *     Things Database.thingsdatabase/main.sqlite
 */

import { join } from "node:path";
import { homedir } from "node:os";
import { existsSync, readdirSync } from "node:fs";
import Database from "better-sqlite3";
import { buildTaskData } from "./taskwarrior.js";
import type { Config, TaskData } from "./types.js";

const THINGS_CONTAINER = join(
  homedir(),
  "Library",
  "Group Containers",
  "JLMPQHK86H.com.culturedcode.ThingsMac",
);

function findThingsDbPath(): string {
  // First try the direct path (older Things versions)
  const directPath = join(THINGS_CONTAINER, "Things Database.thingsdatabase", "main.sqlite");
  if (existsSync(directPath)) return directPath;

  // Search for ThingsData-* subfolder (newer Things versions)
  if (existsSync(THINGS_CONTAINER)) {
    const entries = readdirSync(THINGS_CONTAINER);
    for (const entry of entries) {
      if (entry.startsWith("ThingsData-")) {
        const nestedPath = join(THINGS_CONTAINER, entry, "Things Database.thingsdatabase", "main.sqlite");
        if (existsSync(nestedPath)) return nestedPath;
      }
    }
  }

  throw new Error(
    `Things 3 database not found in ${THINGS_CONTAINER}. ` +
      "Is Things 3 installed on this Mac?",
  );
}

interface ThingsRow {
  uuid: string;
  title: string;
  notes: string | null;
  deadline: string | null;
  project_title: string | null;
  area_title: string | null;
  tags: string | null;
}

/**
 * Query incomplete todos from the Things 3 SQLite database.
 * This mirrors what things.py does internally.
 */
function queryThingsTodos(areaFilter: string[]): ThingsRow[] {
  const dbPath = findThingsDbPath();
  const db = new Database(dbPath, { readonly: true });

  try {
    // TMTask.status: 0 = incomplete, 3 = completed, 2 = cancelled
    // TMTask.trashed: 0 = not trashed
    // TMTask.type: 0 = todo, 1 = project, 2 = heading
    const rows = db
      .prepare(
        `
      SELECT
        t.uuid,
        t.title,
        t.notes,
        CASE WHEN t.deadline IS NOT NULL
          THEN date('2021-01-01', '+' || (t.deadline / 86400) || ' days')
          ELSE NULL
        END AS deadline,
        p.title AS project_title,
        a.title AS area_title,
        GROUP_CONCAT(tag.title, '|||') AS tags
      FROM TMTask t
      LEFT JOIN TMTask p ON t.project = p.uuid
      LEFT JOIN TMArea a ON COALESCE(t.area, p.area) = a.uuid
      LEFT JOIN TMTaskTag tt ON t.uuid = tt.tasks
      LEFT JOIN TMTag tag ON tt.tags = tag.uuid
      WHERE t.status = 0
        AND t.trashed = 0
        AND t.type = 0
      GROUP BY t.uuid
      ORDER BY t.todayIndex
    `,
      )
      .all() as ThingsRow[];

    // Apply area filter if set
    if (areaFilter.length > 0) {
      return rows.filter(
        (r) => r.area_title && areaFilter.includes(r.area_title),
      );
    }

    return rows;
  } finally {
    db.close();
  }
}

export function transformThingsRows(
  rows: ThingsRow[],
  sourceTag: string,
): TaskData[] {
  const result: TaskData[] = [];

  for (const row of rows) {
    // Resolve project name: prefer project, fall back to area
    const project = row.project_title ?? row.area_title ?? undefined;

    // Parse tags (joined with |||)
    const tags = row.tags ? row.tags.split("|||").filter(Boolean) : [];

    // Build annotations from notes
    const annotations: string[] = [];
    if (row.notes) annotations.push(row.notes);

    const taskData = buildTaskData({
      description: row.title,
      externalIdField: "things3_uuid",
      externalId: row.uuid,
      source: "things3",
      sourceTag,
      project,
      due: row.deadline ?? undefined,
      tags,
      annotations,
    });

    result.push(taskData);
  }

  return result;
}

export function fetchThingsTasks(config: Config): TaskData[] {
  const thingsCfg = config.things;
  const syncCfg = config.sync;
  const sourceTag = syncCfg.things_tag;
  const areaFilter = thingsCfg.areas ?? [];

  const rows = queryThingsTodos(areaFilter);
  return transformThingsRows(rows, sourceTag);
}
